packages/artifact/src/internal/requestUtils.ts

```diff
@@ -0,0 +1,74 @@
+import {IHttpClientResponse} from '@actions/http-client/interfaces'
+import {isRetryableStatusCode, isSuccessStatusCode, sleep} from './utils'
+import * as core from '@actions/core'
+
+export async function retry<T>(
+  name: string,
+  method: () => Promise<T>,
+  getStatusCode: (response: T) => number | undefined,
+  errorMessages: Map<number, string>,
+  maxAttempts: number,
+  delay: number
+): Promise<T> {
+  let response: T | undefined = undefined
+  let statusCode: number | undefined = undefined
+  let isRetryable = false
+  let errorMessage = ''
+  let extraErrorInformation: string | undefined = undefined
+  let attempt = 1
+
+  while (attempt <= maxAttempts) {
+    try {
+      response = await method()
+      statusCode = getStatusCode(response)
+
+      if (isSuccessStatusCode(statusCode)) {
+        return response
+      }
+
+      // Extra error information that we want to display if a particular response code is hit
+      if (statusCode) {
+        extraErrorInformation = errorMessages.get(statusCode)
+      }
+
+      isRetryable = isRetryableStatusCode(statusCode)
+      errorMessage = `Artifact service responded with ${statusCode}`
+    } catch (error) {
+      isRetryable = true
+      errorMessage = error.message
+    }
+
+    if (!isRetryable) {
+      core.info(`${name} - Error is not retryable`)
+      break
+    }
+
+    core.info(
+      `${name} - Attempt ${attempt} of ${maxAttempts} failed with error: ${errorMessage}`
+    )
+
+    await sleep(delay)
```

Comment:
Should we increase the delay on each attempt?  I think we should also check `delay > 0`

Reply:
So, the existing methods that are retryable (the PUT calls during upload and GET calls during download, see PR description) have their own retryablity which includes exponential back-off after first checking for a `retry-after` header that we might send back if too many requests are being made. You can see it here:
https://github.com/actions/toolkit/blob/73d5917a6b5ea646ac3173cfceb727ee914ff6ed/packages/artifact/src/internal/upload-http-client.ts#L414-L432

What I did in this PR, is I moved the sleep method to a shared `util` file. The existing methods that are retried are a little bit more complicated because there are multiple calls being done concurrently so they're not switching over to the new `retryHttpClientRequest` method in this PR.  What I am planning on doing in a follow-up PR is:

- Move checking for the `retry-after` header into `util` alongside computing the exponential back-off time there
- Add exponential back-off to `retryHttpClientRequest` that is being added as part of this PR
- Switch over all the HTTP calls to use the same `retryHttpClientRequest` method. 

The last part would require a lot more refactoring and I don't want to make this PR too big so I want to do things in phases. In the past, large updates to the artifact actions would take a while to get through and they were harder to test so I want to split things up into manageable chunks

Reply:
Regardless of what the other code is doing, I think we should still implement exponential backoff in the new code.  The reason we use exponential backoff is that any fixed delay time we choose might be too short and put too much pressure on the service.

Reply:
Added! Most of the code is already there, so it was pretty simple

https://github.com/actions/toolkit/blob/73d5917a6b5ea646ac3173cfceb727ee914ff6ed/packages/artifact/src/internal/utils.ts#L15-L34